xds/src/main/java/io/grpc/xds/internal/certprovider/ZatarCertificateProvider.java

```diff
@@ -0,0 +1,189 @@
+
+  @VisibleForTesting
+  ZatarCertificateProvider(
+      DistributorWatcher watcher,
+      boolean notifyCertUpdates,
+      String directory,
+      String certFile,
+      String privateKeyFile,
+      String trustFile,
+      long refreshIntervalInSeconds,
+      ScheduledExecutorService scheduledExecutorService,
+      TimeProvider timeProvider) {
+    super(watcher, notifyCertUpdates);
+    this.scheduledExecutorService =
+        checkNotNull(scheduledExecutorService, "scheduledExecutorService");
+    this.timeProvider = checkNotNull(timeProvider, "timeProvider");
+    this.directory = Paths.get(checkNotNull(directory, "diretory"));
+    this.certFile = checkNotNull(certFile, "certFile");
+    this.privateKeyFile = checkNotNull(privateKeyFile, "privateKeyFile");
+    this.trustFile = checkNotNull(trustFile, "trustFile");
+    this.refreshIntervalInSeconds = refreshIntervalInSeconds;
+    this.syncContext = createSynchronizationContext(directory);
+  }
+
+  private SynchronizationContext createSynchronizationContext(String details) {
+    final InternalLogId logId = InternalLogId.allocate("ZatarCertificateProvider", details);
+    return new SynchronizationContext(
+        new Thread.UncaughtExceptionHandler() {
+          private boolean panicMode;
+
+          @Override
+          public void uncaughtException(Thread t, Throwable e) {
+            logger.log(
+                Level.SEVERE,
+                "[" + logId + "] Uncaught exception in the SynchronizationContext. Panic!",
+                e);
+            panic(e);
+          }
+
+          void panic(final Throwable t) {
+            if (panicMode) {
+              // Preserve the first panic information
+              return;
+            }
+            panicMode = true;
+            close();
+          }
+        });
+  }
+
+  @Override
+  public void start() {
+    scheduleNextRefreshCertificate(/* delayInSeconds= */0);
+  }
+
+  @Override
+  public void close() {
+    if (scheduledHandle != null) {
+      scheduledHandle.cancel();
+      scheduledHandle = null;
+    }
+    getWatcher().close();
+  }
+
+  private void scheduleNextRefreshCertificate(long delayInSeconds) {
+    if (scheduledHandle != null && scheduledHandle.isPending()) {
+      logger.log(Level.SEVERE, "Pending task found: inconsistent state in scheduledHandle!");
+      scheduledHandle.cancel();
+    }
+    RefreshCertificateTask runnable = new RefreshCertificateTask();
+    scheduledHandle =
+        syncContext.schedule(runnable, delayInSeconds, TimeUnit.SECONDS, scheduledExecutorService);
+  }
+
+  @VisibleForTesting
+  void checkAndReloadCertificates() {
+    try {
+      Path targetPath = Files.readSymbolicLink(directory);
+      if (targetPath.equals(lastModifiedTarget)) {
+        return;
+      }
+      PrivateKey privateKey =
+          CertificateUtils.getPrivateKey(
+              new FileInputStream(new File(targetPath.toFile(), privateKeyFile)));
```

Comment:
Try with resource, or let `CertificateUtils` close the stream (maybe introducing convenient method `CertificateUtils.getPrivateKey(File file) that closes the input stream`)?